File: src/Photokompressor/Core/CompressionEngine.cs
using System.IO;
using ImageMagick;
using NetVips;

namespace Photokompressor.Core;

public enum ResultStatus
{
    Compressed,
    KeptOriginal,
    Failed,
    Cancelled,
}

public record CompressionResult(
    string InputPath,
    string? OutputPath,
    long BeforeBytes,
    long AfterBytes,
    ResultStatus Status,
    string? Note);

public class CompressionEngine
{
    private readonly OutputPathResolver _resolver = new();

    /// <summary>Longest path the native codecs handle reliably; longer paths are bounced through %TEMP%.</summary>
    private const int MaxNativePathLength = 240;

    /// <summary>WebP hard format limit per side.</summary>
    private const int WebpMaxDimension = 16383;

    public static void ConfigureConcurrency(int taskDegreeOfParallelism)
    {
        NetVips.NetVips.Concurrency = Math.Max(2, Environment.ProcessorCount / taskDegreeOfParallelism);

        // libvips' operation cache holds decoded images — and keeps their source files
        // open — which blocks recycling the original in replace mode. Every file is
        // read exactly once here, so the cache buys nothing and only costs us locks.
        Cache.Max = 0;
    }

    public static int TaskDegreeOfParallelism => Math.Clamp(Environment.ProcessorCount / 2, 1, 4);

    public async Task RunBatchAsync(IReadOnlyList<string> files, AppSettings settings,
        IProgress<CompressionResult> progress, CancellationToken ct)
    {
        var frozen = settings.Clone();
        try
        {
            await Parallel.ForEachAsync(files,
                new ParallelOptions { MaxDegreeOfParallelism = TaskDegreeOfParallelism, CancellationToken = ct },
                (file, token) =>
                {
                    progress.Report(CompressFile(file, frozen, token));
                    return ValueTask.CompletedTask;
                });
        }
        catch (OperationCanceledException)
        {
            // Cancelled results for files that never started were already reported
            // by CompressFile's own token check; nothing further to do.
        }
    }

    public CompressionResult CompressFile(string inputPath, AppSettings settings, CancellationToken ct = default,
        bool forceEvenIfLarger = false)
    {
        long beforeBytes = 0;
        string? tempPath = null;
        var cleanup = new List<string>();
        try
        {
            if (ct.IsCancellationRequested)
                return new(inputPath, null, 0, 0, ResultStatus.Cancelled, null);

            var info = new FileInfo(inputPath);
            if (!info.Exists)
                return new(inputPath, null, 0, 0, ResultStatus.Failed, "File not found");
            beforeBytes = info.Length;

            // Replace mode is only safe where Windows will really hand the original
            // back. Checked before any work so nothing is compressed and thrown away.
            if (!settings.KeepOriginals && !SafeReplace.CanRecycle(inputPath, out var blocker))
            {
                return new(inputPath, null, beforeBytes, beforeBytes, ResultStatus.Failed,
                    $"Left alone — {blocker}. Switch on Keep originals to compress it.");
            }

            var finalPath = _resolver.Resolve(inputPath, settings);
            var destDir = Path.GetDirectoryName(finalPath)!;
            Directory.CreateDirectory(destDir);

            tempPath = Path.Combine(destDir, $".pk-tmp-{Guid.NewGuid():N}{settings.ExtensionForFormat()}");
            var nativeTempPath = tempPath;
            if (tempPath.Length > MaxNativePathLength)
            {
                nativeTempPath = Path.Combine(Path.GetTempPath(),
                    $"pk-out-{Guid.NewGuid():N}{settings.ExtensionForFormat()}");
            }

            var nativeInputPath = inputPath;
            if (inputPath.Length > MaxNativePathLength)
            {
                nativeInputPath = Path.Combine(Path.GetTempPath(),
                    $"pk-in-{Guid.NewGuid():N}{Path.GetExtension(inputPath)}");
                File.Copy(inputPath, nativeInputPath, overwrite: true);
                cleanup.Add(nativeInputPath);
            }

            string? note;
            using (var image = LoadPipeline(nativeInputPath, settings, out note))
            {
                Save(image, nativeTempPath, settings);
            }

            if (!ReferenceEquals(nativeTempPath, tempPath) && nativeTempPath != tempPath)
            {
                File.Move(nativeTempPath, tempPath);
            }
            try { File.SetAttributes(tempPath, File.GetAttributes(tempPath) | FileAttributes.Hidden); }
            catch { /* hidden temp is cosmetic */ }

            var afterBytes = new FileInfo(tempPath).Length;
            if (afterBytes <= 0)
                throw new InvalidOperationException("Output file is empty.");

            if (afterBytes >= beforeBytes && !forceEvenIfLarger)
            {
                File.Delete(tempPath);
                return new(inputPath, null, beforeBytes, beforeBytes, ResultStatus.KeptOriginal,
                    "Already smaller — kept original");
            }

            if (settings.KeepOriginals)
            {
                File.SetAttributes(tempPath, FileAttributes.Normal);
                File.Move(tempPath, finalPath, overwrite: false);
            }
            else
            {
                SafeReplace.PromoteTempOverOriginal(tempPath, inputPath, finalPath);
            }
            return new(inputPath, finalPath, beforeBytes, afterBytes, ResultStatus.Compressed, note);
        }
        catch (Exception ex)
        {
            TryDelete(tempPath);
            return new(inputPath, null, beforeBytes, 0, ResultStatus.Failed, FriendlyError(ex));
        }
        finally
        {
            foreach (var path in cleanup)
                TryDelete(path);
        }
    }

    private static Image LoadPipeline(string path, AppSettings settings, out string? note)
    {
        note = null;
        // Size.Down never upscales, so a huge box is a no-op resize that still
        // gives us auto-rotation and ICC handling from the same code path.
        var boxW = settings.ResizeEnabled ? settings.BoxWidth : 5_000_000;
        var boxH = settings.ResizeEnabled ? settings.BoxHeight : 5_000_000;

        Image image;
        if (HeicDecoder.IsHeic(path))
        {
            image = LoadViaMagick(path, settings, boxW, boxH);
        }
        else
        {
            try
            {
                image = Image.Thumbnail(path, boxW, height: boxH, size: Enums.Size.Down, outputProfile: "srgb");
            }
            catch (VipsException)
            {
                // libvips' Windows build lacks some loaders (BMP among them);
                // ImageMagick reads far more, so give it a shot before failing.
                image = LoadViaMagick(path, settings, boxW, boxH);
            }
        }

        try
        {
            if (image.GetTypeOf("n-pages") != 0 && image.Get("n-pages") is int pages && pages > 1)
                note = "Animated input — first frame only";
        }
        catch { /* metadata probe is best-effort */ }

        if (settings.Format == OutputFormat.WebP
            && (image.Width > WebpMaxDimension || image.Height > WebpMaxDimension))
        {
            var shrunk = image.ThumbnailImage(WebpMaxDimension, height: WebpMaxDimension, size: Enums.Size.Down);
            image.Dispose();
            image = shrunk;
            note = note == null ? "Reduced to WebP's 16383 px limit" : note + "; reduced to WebP's 16383 px limit";
        }
        return image;
    }

    private static Image LoadViaMagick(string path, AppSettings settings, int boxW, int boxH)
    {
        var full = HeicDecoder.Load(path);
        if (!settings.ResizeEnabled)
            return full;
        var resized = full.ThumbnailImage(boxW, height: boxH, size: Enums.Size.Down);
        full.Dispose();
        return resized;
    }

    private static void Save(Image image, string tempPath, AppSettings settings)
    {
        switch (settings.Format)
        {
            case OutputFormat.Jpeg:
            {
                var p = Presets.Jpeg(settings.Preset);
                var subsample = p.ForceSubsampleOn ? Enums.ForeignSubsample.On : Enums.ForeignSubsample.Auto;
                var flat = image.HasAlpha() ? image.Flatten(background: [255.0, 255.0, 255.0]) : image;
                try
                {
                    try
                    {
                        flat.Jpegsave(tempPath, q: p.Q, subsampleMode: subsample,
                            optimizeCoding: true, interlace: true, trellisQuant: true,
                            overshootDeringing: true, optimizeScans: true, quantTable: 3,
                            keep: Enums.ForeignKeep.None);
                    }
                    catch (VipsException)
                    {
                        // Bundled libvips without mozjpeg: retry with baseline flags.
                        TryDelete(tempPath);
                        flat.Jpegsave(tempPath, q: p.Q, subsampleMode: subsample,
                            optimizeCoding: true, interlace: true,
                            keep: Enums.ForeignKeep.None);
                    }
                }
                finally
                {
                    if (!ReferenceEquals(flat, image))
                        flat.Dispose();
                }
                break;
            }
            case OutputFormat.Png:
            {
                var p = Presets.Png(settings.Preset);
                if (p.Palette)
                {
                    image.Pngsave(tempPath, compression: 9, palette: true, q: p.Q,
                        dither: p.Dither, effort: p.Effort, keep: Enums.ForeignKeep.None);
                }
                else
                {
                    image.Pngsave(tempPath, compression: 9, keep: Enums.ForeignKeep.None);
                }
                break;
            }
            case OutputFormat.WebP:
            {
                var p = Presets.Webp(settings.Preset);
                image.Webpsave(tempPath, q: p.Q, effort: p.Effort, smartSubsample: true,
                    alphaQ: p.AlphaQ, preset: Enums.ForeignWebpPreset.Photo,
                    keep: Enums.ForeignKeep.None);
                break;
            }
        }
    }

    private static string FriendlyError(Exception ex) => ex switch
    {
        VipsException => "Couldn't read or convert this file",
        MagickException => "Couldn't read this file",
        UnauthorizedAccessException => "Access denied",
        IOException io when (io.HResult & 0xFFFF) == 32 => "File is in use by another program",
        IOException io when (io.HResult & 0xFFFF) == 112 => "Disk full",
        _ => ex.Message,
    };

    private static void TryDelete(string? path)
    {
        if (path == null) return;
        try { if (File.Exists(path)) File.Delete(path); }
        catch { /* cleanup is best-effort */ }
    }
}

File: macos/Sources/PhotokompressorCore/CompressionEngine.swift
import Foundation
import CVips

/// Cancellation flag shared across a batch — the Swift equivalent of
/// CancellationToken, since Foundation has no built-in cooperative token.
public final class CancellationFlag: @unchecked Sendable {
    private let lock = NSLock()
    private var flag = false
    public init() {}
    public func cancel() { lock.lock(); flag = true; lock.unlock() }
    public var isCancelled: Bool { lock.lock(); defer { lock.unlock() }; return flag }
}

private struct VipsOpError: Error { let message: String }

/// Direct port of CompressionEngine.cs: same libvips pipeline, same preset
/// numbers, same "keep original if the result isn't actually smaller", same
/// temp-file-then-atomic-promote write strategy. The two real platform
/// differences are HEIC decode (native ImageIO instead of Magick.NET, see
/// HeicDecoder.swift) and the replace-mode safety check (macOS Trash needs
/// no Windows-style volume/quota preflight, see Trash.swift).
public final class CompressionEngine: @unchecked Sendable {
    private let resolver = OutputPathResolver()

    /// Longest path the native codecs handle reliably; longer paths are bounced through NSTemporaryDirectory().
    private static let maxNativePathLength = 240
    /// WebP hard format limit per side.
    private static let webpMaxDimension: Int32 = 16383

    public init() {}

    public static func configureConcurrency(_ taskDegreeOfParallelism: Int) {
        let cores = ProcessInfo.processInfo.activeProcessorCount
        cvips_set_concurrency(Int32(max(2, cores / max(1, taskDegreeOfParallelism))))
        // libvips' operation cache holds decoded images — and keeps their source files
        // open — which blocks trashing the original in replace mode. Every file is
        // read exactly once here, so the cache buys nothing and only costs us locks.
        cvips_set_cache_max(0)
    }

    public static var taskDegreeOfParallelism: Int {
        min(max(ProcessInfo.processInfo.activeProcessorCount / 2, 1), 4)
    }

    public func runBatch(files: [String], settings: AppSettings, cancellation: CancellationFlag,
                          onResult: @escaping (CompressionResult) -> Void) {
        let frozen = settings
        let queue = DispatchQueue(label: "gr.antidot.photokompressor.compress", attributes: .concurrent)
        let semaphore = DispatchSemaphore(value: Self.taskDegreeOfParallelism)
        let group = DispatchGroup()

        for file in files {
            if cancellation.isCancelled {
                onResult(CompressionResult(inputPath: file, outputPath: nil, beforeBytes: 0, afterBytes: 0,
                                            status: .cancelled, note: nil))
                continue
            }
            semaphore.wait()
            group.enter()
            queue.async { [weak self] in
                defer { semaphore.signal(); group.leave() }
                guard let self else { return }
                if cancellation.isCancelled {
                    onResult(CompressionResult(inputPath: file, outputPath: nil, beforeBytes: 0, afterBytes: 0,
                                                status: .cancelled, note: nil))
                    return
                }
                onResult(self.compressFile(inputPath: file, settings: frozen))
            }
        }
        group.wait()
    }

    public func compressFile(inputPath: String, settings: AppSettings, forceEvenIfLarger: Bool = false) -> CompressionResult {
        let fm = FileManager.default
        var beforeBytes: Int64 = 0
        var tempPath: String?
        var cleanup: [String] = []

        do {
            guard let attrs = try? fm.attributesOfItem(atPath: inputPath),
                  let size = attrs[.size] as? Int64 else {
                return CompressionResult(inputPath: inputPath, outputPath: nil, beforeBytes: 0, afterBytes: 0,
                                          status: .failed, note: "File not found")
            }
            beforeBytes = size

            let finalPath = resolver.resolve(inputPath: inputPath, settings: settings)
            let destDir = (finalPath as NSString).deletingLastPathComponent
            try? fm.createDirectory(atPath: destDir, withIntermediateDirectories: true)

            let ext = settings.extensionForFormat()
            let guid = UUID().uuidString.lowercased().replacingOccurrences(of: "-", with: "")
            let tp = (destDir as NSString).appendingPathComponent(".pk-tmp-\(guid)\(ext)")
            var nativeTempPath = tp
            if tp.utf8.count > Self.maxNativePathLength {
                nativeTempPath = (NSTemporaryDirectory() as NSString).appendingPathComponent("pk-out-\(guid)\(ext)")
            }
            tempPath = tp

            var nativeInputPath = inputPath
            if inputPath.utf8.count > Self.maxNativePathLength {
                let inExt = (inputPath as NSString).pathExtension
                nativeInputPath = (NSTemporaryDirectory() as NSString)
                    .appendingPathComponent("pk-in-\(guid).\(inExt)")
                try fm.copyItem(atPath: inputPath, toPath: nativeInputPath)
                cleanup.append(nativeInputPath)
            }

            var note: String?
            let image = try loadPipeline(path: nativeInputPath, settings: settings, note: &note)
            defer { VipsRuntime.unref(image) }
            try save(image: image, path: nativeTempPath, settings: settings)

            if nativeTempPath != tp {
                try fm.moveItem(atPath: nativeTempPath, toPath: tp)
            }

            let afterAttrs = try fm.attributesOfItem(atPath: tp)
            let afterBytes = (afterAttrs[.size] as? Int64) ?? 0
            if afterBytes <= 0 {
                throw VipsOpError(message: "Output file is empty.")
            }

            if afterBytes >= beforeBytes && !forceEvenIfLarger {
                try? fm.removeItem(atPath: tp)
                return CompressionResult(inputPath: inputPath, outputPath: nil, beforeBytes: beforeBytes,
                                          afterBytes: beforeBytes, status: .keptOriginal,
                                          note: "Already smaller — kept original")
            }

            if settings.keepOriginals {
                try fm.moveItem(atPath: tp, toPath: finalPath)
            } else {
                try Trash.promoteTempOverOriginal(tempPath: tp, originalPath: inputPath, finalPath: finalPath)
            }
            for path in cleanup { try? fm.removeItem(atPath: path) }
            return CompressionResult(inputPath: inputPath, outputPath: finalPath, beforeBytes: beforeBytes,
                                      afterBytes: afterBytes, status: .compressed, note: note)
        } catch {
            if let tp = tempPath { try? fm.removeItem(atPath: tp) }
            for path in cleanup { try? fm.removeItem(atPath: path) }
            return CompressionResult(inputPath: inputPath, outputPath: nil, beforeBytes: beforeBytes, afterBytes: 0,
                                      status: .failed, note: Self.friendlyError(error))
        }
    }

    // MARK: - Pipeline

    private func loadPipeline(path: String, settings: AppSettings, note: inout String?) throws -> VipsImageRef {
        note = nil
        // A huge box is a no-op resize (never upscales) that still gives us
        // auto-rotation and ICC handling from the same code path.
        let boxW: Int32 = settings.resizeEnabled ? Int32(settings.boxWidth) : 5_000_000
        let boxH: Int32 = settings.resizeEnabled ? Int32(settings.boxHeight) : 5_000_000

        var image: VipsImageRef
        if HeicDecoder.isHeic(path: path) {
            image = try loadViaImageIO(path: path, settings: settings, boxW: boxW, boxH: boxH)
        } else if let vipsLoaded = cvips_thumbnail_file(path, boxW, boxH) {
            image = vipsLoaded
        } else {
            // The bundled libvips doesn't cover every format (e.g. some BMP
            // variants); ImageIO reads far more, so give it a shot before failing.
            image = try loadViaImageIO(path: path, settings: settings, boxW: boxW, boxH: boxH)
        }

        let pages = cvips_get_n_pages(image)
        if pages > 1 {
            note = "Animated input — first frame only"
        }

        if settings.format == .webP {
            let w = cvips_image_get_width(image)
            let h = cvips_image_get_height(image)
            if w > Self.webpMaxDimension || h > Self.webpMaxDimension {
                guard let shrunk = cvips_thumbnail_image(image, Self.webpMaxDimension, Self.webpMaxDimension) else {
                    throw VipsOpError(message: VipsRuntime.lastError)
                }
                VipsRuntime.unref(image)
                image = shrunk
                note = note == nil ? "Reduced to WebP's 16383 px limit" : note! + "; reduced to WebP's 16383 px limit"
            }
        }
        return image
    }

    private func loadViaImageIO(path: String, settings: AppSettings, boxW: Int32, boxH: Int32) throws -> VipsImageRef {
        let full = try HeicDecoder.load(path: path)
        if !settings.resizeEnabled {
            return full
        }
        guard let resized = cvips_thumbnail_image(full, boxW, boxH) else {
            VipsRuntime.unref(full)
            throw VipsOpError(message: VipsRuntime.lastError)
        }
        VipsRuntime.unref(full)
        return resized
    }

    private func save(image: VipsImageRef, path: String, settings: AppSettings) throws {
        switch settings.format {
        case .jpeg:
            let p = Presets.jpeg(settings.preset)
            let hasAlpha = cvips_has_alpha(image) != 0
            let flat: VipsImageRef
            if hasAlpha {
                guard let flattened = cvips_flatten_white(image) else {
                    throw VipsOpError(message: VipsRuntime.lastError)
                }
                flat = flattened
            } else {
                flat = image
            }
            defer { if flat != image { VipsRuntime.unref(flat) } }

            let subsampleOn: Int32 = p.forceSubsampleOn ? 1 : 0
            if cvips_jpegsave_mozjpeg(flat, path, p.q, subsampleOn) != 0 {
                // Bundled libvips without mozjpeg support: retry with baseline flags.
                try? FileManager.default.removeItem(atPath: path)
                if cvips_jpegsave_baseline(flat, path, p.q, subsampleOn) != 0 {
                    throw VipsOpError(message: VipsRuntime.lastError)
                }
            }
        case .png:
            let p = Presets.png(settings.preset)
            let rc: Int32
            if p.palette {
                rc = cvips_pngsave_palette(image, path, p.q, p.dither, p.effort)
            } else {
                rc = cvips_pngsave_lossless(image, path)
            }
            if rc != 0 { throw VipsOpError(message: VipsRuntime.lastError) }
        case .webP:
            let p = Presets.webp(settings.preset)
            if cvips_webpsave(image, path, p.q, p.effort, p.alphaQ) != 0 {
                throw VipsOpError(message: VipsRuntime.lastError)
            }
        }
    }

    // MARK: - Errors

    private static func friendlyError(_ error: Error) -> String {
        if let vipsError = error as? VipsOpError {
            return vipsError.message.isEmpty ? "Couldn't read or convert this file" : vipsError.message
        }
        if let heicError = error as? HeicDecoder.HeicError {
            switch heicError {
            case .decodeFailed, .renderFailed: return "Couldn't read this file"
            case .vipsFailed: return "Couldn't convert this file"
            }
        }
        let nsError = error as NSError
        if nsError.domain == NSCocoaErrorDomain {
            switch nsError.code {
            case NSFileWriteNoPermissionError, NSFileReadNoPermissionError: return "Access denied"
            case NSFileWriteOutOfSpaceError: return "Disk full"
            default: break
            }
        }
        if nsError.domain == NSPOSIXErrorDomain {
            switch nsError.code {
            case Int(EACCES): return "Access denied"
            case Int(ENOSPC): return "Disk full"
            case Int(EBUSY), Int(ETXTBSY): return "File is in use by another program"
            default: break
            }
        }
        return nsError.localizedDescription
    }
}
